File: dotnet/src/Turkcell.BT.Dotnet.Lib/BeyondTrustAuthParsing.cs
namespace Turkcell.BT.Dotnet.Lib;

internal sealed record ParsedApiKey(string Key, string? RunAsUser)
{
    public string ToAuthorizationHeader()
    {
        return string.IsNullOrWhiteSpace(RunAsUser)
            ? $"PS-Auth key={Key};"
            : $"PS-Auth key={Key}; runas={RunAsUser};";
    }
}

internal static class BeyondTrustAuthParsing
{
    public static bool TryParseApiKey(string? rawValue, string? explicitRunAsUser, out ParsedApiKey? parsedApiKey)
    {
        var candidate = (rawValue ?? string.Empty).Trim();
        if (candidate.StartsWith("PS-Auth", StringComparison.OrdinalIgnoreCase))
        {
            candidate = candidate["PS-Auth".Length..].Trim();
        }

        string? key = null;
        string? inlineRunAs = null;

        foreach (var part in candidate.Split(';', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries))
        {
            if (part.StartsWith("key=", StringComparison.OrdinalIgnoreCase))
            {
                key = part[4..].Trim();
            }
            else if (part.StartsWith("runas=", StringComparison.OrdinalIgnoreCase))
            {
                inlineRunAs = part[6..].Trim();
            }
            else if (string.IsNullOrWhiteSpace(key))
            {
                key = part.Trim();
            }
        }

        if (string.IsNullOrWhiteSpace(key))
        {
            parsedApiKey = null;
            return false;
        }

        var runAsUser = !string.IsNullOrWhiteSpace(explicitRunAsUser)
            ? explicitRunAsUser.Trim()
            : inlineRunAs;

        parsedApiKey = new ParsedApiKey(
            key.Trim(),
            string.IsNullOrWhiteSpace(runAsUser) ? null : runAsUser.Trim());

        return true;
    }
}

File: java/turkcell-bt-java-lib/src/main/java/com/turkcell/bt/java/BeyondTrustAuthParsing.java
package com.turkcell.bt.java;

final class BeyondTrustAuthParsing {
    private BeyondTrustAuthParsing() {
    }

    static ParsedApiKey parseApiKey(String rawValue, String explicitRunAsUser) {
        String candidate = rawValue == null ? "" : rawValue.trim();
        if (candidate.regionMatches(true, 0, "PS-Auth", 0, "PS-Auth".length())) {
            candidate = candidate.substring("PS-Auth".length()).trim();
        }

        String key = null;
        String inlineRunAs = null;

        for (String part : candidate.split(";")) {
            String trimmed = part.trim();
            if (trimmed.isEmpty()) {
                continue;
            }

            if (trimmed.regionMatches(true, 0, "key=", 0, 4)) {
                key = trimmed.substring(4).trim();
            } else if (trimmed.regionMatches(true, 0, "runas=", 0, 6)) {
                inlineRunAs = trimmed.substring(6).trim();
            } else if (key == null || key.isBlank()) {
                key = trimmed;
            }
        }

        if (key == null || key.isBlank()) {
            return null;
        }

        String runAsUser = explicitRunAsUser != null && !explicitRunAsUser.isBlank()
                ? explicitRunAsUser.trim()
                : inlineRunAs;

        return new ParsedApiKey(key.trim(), runAsUser == null || runAsUser.isBlank() ? null : runAsUser.trim());
    }

    static final class ParsedApiKey {
        private final String key;
        private final String runAsUser;

        ParsedApiKey(String key, String runAsUser) {
            this.key = key;
            this.runAsUser = runAsUser;
        }

        String toAuthorizationHeader() {
            return runAsUser == null || runAsUser.isBlank()
                    ? "PS-Auth key=" + key + ";"
                    : "PS-Auth key=" + key + "; runas=" + runAsUser + ";";
        }
    }
}
